csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,361 @@
+      using var which = await Task.WhenAny(front ?? next,
+                                           next)
+                                  .ConfigureAwait(false);
+
+      // If there is an error or cancellation has been requested, we must throw the exception
+      which.ThrowIfError(cts);
+
+      // semaphore has been acquired so we can enqueue a new task
+      if (ReferenceEquals(which,
+                          semAcquire))
+      {
+        move = Task.Run(MoveNext,
+                        cancellationToken);
+        // prepare the new semaphore acquisition
+        semAcquire = sem.WaitAsync(cancellationToken);
+        // The next thing to do would now be to move to the next iteration
+        next = move;
+        continue;
+      }
+
+      // Front task was ready, so we can get its result and yield it
+      if (ReferenceEquals(which,
+                          front))
+      {
+        yield return await queue.Dequeue()
+                                .ConfigureAwait(false);
+        continue;
+      }
+
+      // Move has completed ("next iteration")
+      Debug.Assert(ReferenceEquals(which,
+                                   move));
+
+      // Iteration has reached an end
+      if (!await move.ConfigureAwait(false))
+      {
+        break;
+      }
+
+      // Current task is now ready,
+      // so now we can enqueue it as soon as the semaphore is acquired
+      var x = enumerator.Current;
+
+      // Async closure that waits for the task x and releases the semaphore
+      async Task<T> TaskLambda()
+      {
+        var res = await x.ConfigureAwait(false);
+        sem.Release();
+        return res;
+      }
+
+      // We can enqueue the task
+      queue.Enqueue(Task.Run(TaskLambda,
+                             cancellationToken));
+
+      next = semAcquire;
+    }
+
+    // We finished iterating over the inputs and
+    // must now wait for all the tasks in the queue
+    foreach (var task in queue)
+    {
+      // Dequeue a new task and wait for its result
+      // As the task is wrapped in a Task.run with the cancellation token,
+      // It will be cancelled as soon cancellation is triggered
+      using (task)
+      {
+        yield return await task.ConfigureAwait(false);
+      }
+    }
+  }
+
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks that call `func`.
+  ///   The maximum number of tasks in flight at any given moment is given in the `parallelTaskOptions`.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelTaskOptions">Options (eg: parallelismLimit, cancellationToken)</param>
+  /// <param name="func">Function to spawn on the enumerable input</param>
+  /// <typeparam name="TU">Type of the inputs</typeparam>
+  /// <typeparam name="TV">Type of the outputs</typeparam>
```

Comment:
```suggestion
  /// <typeparam name="TInput">Type of the inputs</typeparam>
  /// <typeparam name="TOuput">Type of the outputs</typeparam>
```
Way clearer this way